src/linker/Linker/UnconditionalSuppressMessageAttributeState.cs

```diff
@@ -0,0 +1,244 @@
+	{
+		private readonly AssemblyDefinition _assembly;
+		private GlobalSuppressions _lazyGlobalSuppressions;
+		private readonly Dictionary<MetadataToken, ImmutableDictionary<string, SuppressMessageInfo>> _localSuppressionsByMdToken;
+
+		private static readonly Dictionary<string, TargetScope> _suppressMessageScopeTypes = new Dictionary<string, TargetScope> (StringComparer.OrdinalIgnoreCase)
+		{
+			{ string.Empty, TargetScope.None },
+			{ "module", TargetScope.Module },
+			{ "namespace", TargetScope.Namespace },
+			{ "resource", TargetScope.Resource },
+			{ "type", TargetScope.Type },
+			{ "member", TargetScope.Member },
+			{ "namespaceanddescendants", TargetScope.NamespaceAndDescendants }
+		};
+
+		private static bool TryGetTargetScope (SuppressMessageInfo info, out TargetScope scope)
+			=> _suppressMessageScopeTypes.TryGetValue (info.Scope ?? string.Empty, out scope);
+
+		private bool HasLocalSuppressions { get; set; }
+		private bool HasGlobalSuppressions { get; set; }
+		public bool HasSuppressions {
+			get {
+				return (HasLocalSuppressions || HasGlobalSuppressions);
+			}
+		}
+
+		public UnconditionalSuppressMessageAttributeState (AssemblyDefinition assembly)
+		{
+			_assembly = assembly;
+			_localSuppressionsByMdToken = new Dictionary<MetadataToken, ImmutableDictionary<string, SuppressMessageInfo>> ();
+		}
+
+		public void AddLocalSuppression (CustomAttribute ca, MetadataToken mdToken)
+		{
+			SuppressMessageInfo info;
+			if (!TryDecodeSuppressMessageAttributeData (ca, out info)) {
+				return;
+			}
+
+			if (_localSuppressionsByMdToken.ContainsKey (mdToken)) {
+				_localSuppressionsByMdToken[mdToken].Add (info.Id, info);
+			} else {
+				_localSuppressionsByMdToken[mdToken] = new Dictionary<string, SuppressMessageInfo> { { info.Id, info } }.ToImmutableDictionary ();
+			}
+
+			HasLocalSuppressions = true;
+		}
+
+		public bool IsSuppressed (string id, MessageOrigin warningOrigin, out SuppressMessageInfo info)
+		{
+			info = default;
+
+			if (HasGlobalSuppressions && IsGloballySuppressed (id, out info)) {
+				return true;
+			}
+
+			if (HasSuppressions && warningOrigin.MdTokenProvider != null) {
+				IMetadataTokenProvider mdTokenProvider = warningOrigin.MdTokenProvider;
+				do {
+					if (IsLocallySuppressed (id, mdTokenProvider.MetadataToken, out info)) {
+						return true;
+					}
+
+					if (mdTokenProvider is FieldDefinition ||
+						mdTokenProvider is PropertyDefinition ||
+						mdTokenProvider is EventDefinition ||
+						mdTokenProvider is MethodDefinition) {
+						mdTokenProvider = (mdTokenProvider as IMemberDefinition).DeclaringType;
+					} else if (mdTokenProvider is TypeDefinition type) {
+						if (type.IsNested) {
+							_ = type.BaseType.Resolve ();
+						}
+
+						mdTokenProvider = type.Module;
+					} else if (mdTokenProvider is ModuleDefinition module) {
+						mdTokenProvider = null;
+					}
+				} while (mdTokenProvider != null);
```

Comment:
Does this loop make sense when we care only about members and declaration?

Reply:
I've rewritten this, we now simply walk up members up to the declaring type to check for suppressions.